hack/verify-prometheus-imports.sh

```diff
@@ -0,0 +1,102 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script validates that only a restricted set of packages are importing
+# github.com/prometheus/*
+
+# NOTE: this is not the same as verify-imports which can only verify
+# that within a particular package the imports made are allowed.
+#
+# This is also not the same thing as verify-import-boss, which is pretty
+# powerful for specifying restricted imports but does not scale to checking
+# the entire source tree well and is only enabled for specific packages.
+#
+# See: https://github.com/kubernetes/kubernetes/issues/99876
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+source "${KUBE_ROOT}/hack/lib/init.sh"
+source "${KUBE_ROOT}/hack/lib/util.sh"
+
+# See: https://github.com/kubernetes/kubernetes/issues/89267
+allowed_prometheus_importers=(
+  ./cluster/images/etcd-version-monitor
+  ./pkg/volume/util/operationexecutor
+  ./staging/src/k8s.io/component-base/metrics
+  ./staging/src/k8s.io/component-base/metrics/legacyregistry
+  ./staging/src/k8s.io/component-base/metrics/testutil
+  ./test/e2e/apimachinery
+  ./test/e2e_node
+  ./test/integration/apiserver/flowcontrol
+  ./test/integration/metrics
+)
+
+# Go imports always involve a double quoted string of the package path
+# https://golang.org/ref/spec#Import_declarations
+# We can cheaply locate possible importers with a simple regex / recursive grep
+# Afterwards we post-process to verify that it's not an errant string literal
+# more cheaply than actually loading the entire import graph for the repo.
+packages_probably_importing_prometheus=()
+while IFS='' read -r line; do
+  # convert from file to package, and only insert unique results
+  # we want to minimize the amount of `go list` calls we need to make
+  pkg="$(dirname "${line}")"
+  if ! kube::util::array_contains "$pkg" "${packages_probably_importing_prometheus[@]}"; then
+    packages_probably_importing_prometheus+=("$pkg")
+  fi
+done < <(grep \
+  --exclude-dir={_output,vendor} \
+  --include='*.go' \
+  -R . \
+  -l \
+  -Ee '"github.com/prometheus/.*"' \
+| LC_ALL=C sort -u)
+
+# check each possible package against `go list` to verify that it really is
+# an import and not an errant string
+really_failing=()
+for pkg in "${packages_probably_importing_prometheus[@]}"; do
+  if go list \
+        -f '{{ join .Imports "\n" }}{{ printf "\n" }}{{ join .TestImports "\n" }}' \
+        "$pkg" \
+    | grep -E \
+        '^(k8s.io/kubernetes/vendor/)?github\.com/prometheus/.*$' \
```

Comment:
`(k8s.io/kubernetes/vendor/)?` is because:

```console
$ GO111MODULE=off go list -f '{{ join .Imports "\n" }}{{ printf "\n" }}{{ join .TestImports "\n" }}' ./cluster/images/etcd-version-monitor | grep  'github\.com/prometheus/.*'
k8s.io/kubernetes/vendor/github.com/prometheus/client_model/go

$ GO111MODULE=on go list -f '{{ join .Imports "\n" }}{{ printf "\n" }}{{ join .TestImports "\n" }}' ./cluster/images/etcd-version-monitor | grep  'github\.com/prometheus/.*'
github.com/prometheus/client_model/go
```

Currently in our scripts we have the former case, but we will need to move to the latter someday most likely.

Reply:
This also seems like it could be a comment in file (sans invocations, just an explanation that we want to be future-proof for module support)

Reply:
agreed, but this code is now dropped.